Introduce search context - point in time view of indices

dnhatn opened the pull request:
#### Open a new search context:

```
POST /twitter/_search_context?keep_alive=5m
```

#### Execute a search with a specific search context

```
POST /_search
{
    "size": 100,
    "query": {
        "match" : {
            "title" : "elasticsearch"
        }
    },
    "search_context": {
       "id":  "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB",
       "keep_alive": "5m" 
    }
}
```

#### Close a search context when it's no longer needed
```
DELETE /_search_context
{
    "id" : "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB"
}
```

#### Notable works in this change:

- Move the search state to the coordinating node: https://github.com/elastic/elasticsearch/pull/52741
- Allow searches with a specific reader context: https://github.com/elastic/elasticsearch/pull/53989
- Adds the ability to acquire readers in IndexShard: https://github.com/elastic/elasticsearch/pull/54966

Relates #46523
Relates #26472


Co-authored-by: Jim Ferenczi <jimczi@apache.org>

dnhatn committed:
Cut over from SearchContext to ReaderContext (#51282)

With this change, we partially move the state of SearchContext to 
ReaderContext. This is another step allowing us to move the state of
search to the coordinating node.

We will need several follow-ups to move the entire search state to the 
coordinating node.

Relates #46523

dnhatn committed:
Cut over from SearchContext to ReaderContext (#51282)

With this change, we partially move the state of SearchContext to
ReaderContext. This is another step allowing us to move the state of
search to the coordinating node.

We will need several follow-ups to move the entire search state to the
coordinating node.

Relates #46523

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Revert "Cut over from SearchContext to ReaderContext (#51282)"

This reverts commit 3ce9a3bf9663e83636af3cc30072f3a9dfa38692.

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Move states of search to coordinating node (#52741)

This commit moves the states of search to the coordinating node instead 
of keeping them in the data node. 

Relates #46523

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Adjust SearchService after merging from master

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Allow searches with specific reader contexts (#53989)

jimczi committed:
Merge branch 'master' into feature/reader-context

jimczi committed:
fix checkstyle after backport

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Adjust hlrc tests after merge

dnhatn committed:
enable bwc

dnhatn committed:
Restore missing rewrite when create search context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Add open reader contexts API (#55265)

This commit adds a new API that allows opening reader contexts 
in a separate step before using them in subsequent search requests

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

jimczi committed:
Adds the ability to acquire readers in IndexShard (#54966)

This change adds the ability to acquire a point in time reader on an engine.
This is needed for frozen indices that lazily loads the reader on every
phase of the search requests. Acquiring a reader on a frozen index ensures
that the engine will not be closed until the reader is released, leaving the
directory reader unopened until a call to acquire a searcher is made.
When the searcher is closed, the underlyinng directory reader is also closed
unless another requests on the same frozen shard is in-flight. This ensures
that the directory reader of frozen indices is opened only when requests are
executed (they consume a thread in the search throttled pool).

jimczi committed:
Mark the reader context as used (#55854)

This commit ensures that we protect the reader context
from being removed during the full lifecycle of a search phase.

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Enable can match for search with reader contexts (#56032)

Enabling the can-match phase for search requests with reader contexts

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Move SearchWithReaderContextIT to internalClusterTest

dnhatn committed:
Rename reader context to search context (#56351)

This change renames reader context to search context in the rest layer.

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Add total hits info to test assertion

dnhatn committed:
more on renaming

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
fix security listener

dnhatn committed:
simplify validation of reader context

dnhatn committed:
more on renaming

dnhatn committed:
bwc: remove scroll on failed authorization

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn commented:
@elasticmachine test this please

dnhatn commented:
@elasticmachine test this please

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Remove reader context if index gets index midway

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

mayya-sharipova commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -0,0 +1,92 @@
+    }
+}
+
+<1> A search request with `search_context` must not specify `index`, `routing`,
```
Does it mean that this search request will be forwarded to specific shards?  
If yes, what happens if these shards are no longer avaialable?

dnhatn replied:
Thanks for checking, Mayya. The request will fail if some shards are not available (depending on the partial result flag). The scroll API has the same limitation. However, as the search context becomes stateless, we will work a follow-up to replace the unavailable shards with other copies when possible.

dnhatn committed:
Merge branch 'master' into feature/reader-context

jpountz commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -621,35 +608,64 @@ public Searcher acquireSearcher(String source, SearcherScope scope) throws Engin
+                        /* In general, readers should never be released twice or this would break reference counting. There is one rare case
+                         * when it might happen though: when the request and the Reaper thread would both try to release it in a very short
+                         * amount of time, this is why we only log a warning instead of throwing an exception. */
+                        logger.warn("Reader was released twice", new IllegalStateException("Double release"));
```
should we add an `assert false` too to make sure we'd catch it in tests?

jimczi replied:
the comment above explains why we don't add an assert or throw an exception. The race condition can still happen.

dnhatn replied:
I think this is no longer an issue as we are using refCounted for ReaderContext. I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/44aca91af592aad6ddee482019091a2a8b8ee90a.

jpountz reviewed (commented):
Very exciting feature! I'm not done with the review (this is a large change :) ) but it made me wonder about how we should refer to this feature. I've read this feature referred to as search contexts, reader context and point-in-time views, we should stick to one and try to use it consistently?

dnhatn committed:
Merge branch 'master' into feature/reader-context

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -342,6 +381,28 @@ public void canReturnNullResponseIfMatchNoDocs(boolean value) {
+     * Returns <code>true</code> if the request needs to create a long-lived context.
+     */
+    public boolean shouldCreatePersistentReader() {
+        return scroll != null  || (keepAlive != null && readerId == null);
```
Should it be `readerId != null`?

jimczi replied:
This was needed when the search context was created on the first search. We have a dedicated API to create search context now so this condition is never met (readerId cannot be null). I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/d654b263970c570677c951fbdfb12018b9311019.

jimczi committed:
remove unused code

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -333,20 +340,18 @@ public void onFailure(Exception exc) {
+                                            boolean keepStatesInContext) throws IOException {
+        ReaderContext readerContext = createOrGetReaderContext(request, keepStatesInContext);
+        try (Releasable ignored = readerContext.markAsUsed();
+             SearchContext context = createContext(readerContext, request, task, true)) {
```
nit: ident is wrong

dnhatn committed:
Merge branch 'master' into feature/reader-context

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -550,136 +562,194 @@ private Executor getExecutor(IndexShard indexShard) {
+        final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        final IndexShard shard = indexService.getShard(shardId.id());
+        final SearchOperationListener searchOperationListener = shard.getSearchOperationListener();
+        shard.awaitShardSearchActive(ignored -> {
```
I am not very clear on what thread pool we are going to execute this code after a shard has been refreshed?  Refresh thread pool?

dnhatn replied:
If the shard is not search-idle, then this will be called immediately using the network thread pool; otherwise, it will be called using the refresh thread pool.

mayya-sharipova replied:
Thanks for clarifying, Nhat

Mpdreamz commented:
I wonder if we can extend `scroll` instead? 

e.g 
```json
POST /twitter/_search?point_in_time_scroll=1m
{
    "size": 100,
    "query": {
        "match" : {
            "title" : "elasticsearch"
        }
    }
}
```

The returned id encodes `search_after` values automatically based on the sort fields.

```json
POST /_search/scroll 
{
    "scroll" : "1m", 
    "scroll_id" : "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB" 
}
```

`_search/scroll` knows the scroll is using a search context. Using `point_in_time_scroll` the `scroll_id` is always changing and the same `scroll_id` can be reused.


```json
DELETE /_search/scroll 
{
    "scroll_id" : "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB" 
}
```
Could unpack any `scroll_id` to clear the right context.

Not sure if `slice` can be supported using search_context?

Reason I raise this is too limit of number of specialized API's and the fact it would allow an easy path for folks already using scroll.

I realize I am coming in hot and cold at the same time, but curious about the caveats of this approach. 


The API as it stands LGTM, being able to retry is huge! 

A `body` with `DELETE` is rare, currently only the `Clear Scroll API` supports this. It is allowed by the HTTP specification however but [often http abstractions do not support it](https://github.com/angular/angular/issues/19438).

Since it allowed by the spec and has prior art lets first roll with the implementation as is before exploring adding `DELETE /_search_context/{id}`. `scroll_id` was likely to exceed request PATH max length limits, not sure if `_search_context` encodes the same possible amount of information.

Bike-shedding nitpick:

I think `search_context` is too technical and ambiguous with `filter context` and `query context` being very established terms already. 

`_search/_freeze` or `_search/_snapshot` are equally ambiguous with other features. `_search/_point_in_time` is very underscore heavy. Don't really have an alternative `</bikeshed>`

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn commented:
@Mpdreamz Thank you for reviewing.

Yeah, we also considered wrapping and making scroll retry-able (see [TODO](https://github.com/elastic/elasticsearch/blob/3e39e8097c2f54f2b3f860b213cc672710b03d50/server/src/main/java/org/elasticsearch/action/search/SearchScrollAsyncAction.java#L56-L70) and https://github.com/elastic/elasticsearch/issues/56828). However, we prefer to deprecate the scroll API and encourage users to use search_after with search context explicitly. @jimczi Can you chime in our reason here?

dnhatn committed:
Use IndexShard from reader context (#57384)

Previously, we resolve IndexService and IndexShard once when creating a 
search context. However, since #51282, we resolve them in every phase as
we recreate a search context in each phase. This change can cause the
dfs_query or fetch phase to fail with ShardNotFoundException while
previously did not. This behavior is quite subtle, but I think we should
make it in line with the master branch.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/internal/ReaderContext.java:
```diff
@@ -0,0 +1,203 @@
+        return null;
+    }
+
+    public AggregatedDfs getAggregatedDfs(AggregatedDfs other) {
```
Is it worth to remove `getAggregatedDfs` and `setAggregatedDfs` from  `ReaderContext` and have them only in `LegacyReaderContext`?

dnhatn replied:
Yeah, this was the first choice. However, I went with a different approach to avoid too much casting.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/RescoreDocIds.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+import java.util.Set;
+
+public final class RescoreDocIds implements Writeable {
```
Would be nice to add java documentation describing why is the purpose of this class?

 Is the goal if some shards fail between querying and fetching phase, a coordinating node can continue fetching phase using other shards, including explanation of rescoring where this class would be necessary?

dnhatn replied:
I've added a javadocs in https://github.com/elastic/elasticsearch/pull/56480/commits/aeeee70b15305fa9a7450f25a3efe1a8a3216bba.

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -714,4 +777,38 @@ private void maybeFinish() {
+                if (shard.currentNodeId().equals(entry.getValue().getNode())) {
+                    matchingNodeFirstRoutings.add(0, shard);
+                } else {
+                    matchingNodeFirstRoutings.add(shard);
```
why do we use shards from other nodes?  Is it that they will not have matching searchContext any way?

dnhatn replied:
We discussed this offline. We are adding every shard copy to the iterator for these two reasons:

- Reuse the existing error handling that we have in TransportSearchAction
- We can plugin the retry logic if the target shards are not available

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -496,7 +552,7 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
         }
         if (searchRequest.allowPartialSearchResults() == null) {
-           // No user preference defined in search request - apply cluster service default
+            // No user preference defined in search request - apply cluster service default
```
We have already set this on line 500, do we need this again here?

dnhatn replied:
Addressed in https://github.com/elastic/elasticsearch/commit/0bc08e15223251b4ddb2d099f3a17222bf991a9f.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1109,6 +1136,9 @@ public void parseXContent(XContentParser parser, boolean checkTrailingTokens) th
                     collapse = CollapseBuilder.fromXContent(parser);
+                } else if (SEARCH_CONTEXT.match(currentFieldName, parser.getDeprecationHandler())) {
+                    searchContextBuilder = SearchContextBuilder.
+                        fromXContent(parser);
```
nit: should lines 1140 and 1141 be combined?

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -465,26 +491,56 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
+            assert searchRequest.searchContextBuilder() != null;
+            aliasFilter = new HashMap<>();
+            for (ShardId shardId : searchContexts.keySet()) {
+                aliasFilter.put(shardId.getIndex().getUUID(), AliasFilter.EMPTY);
```
Do we expect a search request with `search_context` work with filtered alias?  If not, may be we should explain in the `search_context.asciidoc` that users should not open search contexts on filtered aliases.
Because currently  we can use filtered alias in an open search context request, but if we use this search context in search requests, the search results will not be filtered.

jimczi replied:
Good catch, this is not implemented. We should add the filtered aliases in the reader context, I'll try to tackle this next week.

jimczi replied:
Implemented in https://github.com/elastic/elasticsearch/pull/56480/commits/96fe51f860d5aa5184aad53bbf7101b8d5671b0f

ywelsch reviewed (commented):
I've looked only at Engine and friends, and left some minor questions regarding exceptional control flow, o.w. this looks mostly good to me.

jimczi committed:
Update docs/reference/search/search_context.asciidoc

Co-authored-by: Adrien Grand <jpountz@gmail.com>

jimczi committed:
Update docs/reference/search/search_context.asciidoc

Co-authored-by: Adrien Grand <jpountz@gmail.com>

mayya-sharipova commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -0,0 +1,92 @@
+    "id" : "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB"
+}
+---------------------------------------
+// TEST[catch:missing]
```
may be also put a sample response and describe it, something like:
```js
{
    "succeeded": true,
    "num_freed": 3
}
```
num_freed  corresponds to the number of shards on which contexts were deleted.

dnhatn replied:
Added in https://github.com/elastic/elasticsearch/pull/56480/commits/1d9bb74f0a303313887b66b6f751c2bc9bd8ce87.

jimczi committed:
address review

mayya-sharipova commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -0,0 +1,92 @@
+the time to live of the search context
+
+IMPORTANT: The open search context request and each subsequent search request can
+return different `id`; thus always use the most recently received `id` for the
```
I am interested in what situations search context id can be different between requests?  I assume if all the original shards are still available, we expect search context is to be the same?

jimczi replied:
I think we should remove this disclaimer. We don't change the `reader_id` today but I think this was added to copy scrolls which have the same warning. That leaves the door open in case we want to change the scroll id in the future. However this is different than scrolls here since the `reader_id` can be shared so I don't think this should be left open.
@dnhatn wdyt ?

dnhatn replied:
Yeah, that was my intention. When we implement the retry logic (i.e., try with other shards when the target shards are not available), the id will change. I am okay to remove it now if it's confusing.

jimczi committed:
Merge branch 'master' into reader-context

jimczi commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -0,0 +1,92 @@
+	    "id":  "46ToAwEHdHdpdHRlchZSa1dRZDFELVNLbTZkQVJudUtGMFFnAAAWNVAwS09JWTdTRUdibWE1ZC1id0tjQRRURnhPWjNFQmI3bmVOeVZReS1tRAAAAAAAAAAB", <2>
+	    "keep_alive": "1m"  <3>
+    }
+}
```
I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/0bc08e15223251b4ddb2d099f3a17222bf991a9f

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -316,14 +319,18 @@ protected void doClose() {
-                runAsync(shard, () -> executeDfsPhase(request, task), listener);
+                try {
+                    runAsync(shard, () -> executeDfsPhase(request, task, keepStatesInContext), listener);
+                } catch (Exception exc) {
```
This is confusing indeed. I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/0bc08e15223251b4ddb2d099f3a17222bf991a9f to have a single `runAsync` method that does the right thing.

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -363,37 +368,61 @@ private void loadOrExecuteQueryPhase(final ShardSearchRequest request, final Sea
+                        if (canRewriteToMatchNone(canMatchRequest.source())
+                                && canMatchRequest.source().query() instanceof MatchNoneQueryBuilder) {
+                            if (orig.readerId() != null) {
+                                try (releasable) {
```
That's consistent with the other part of this class that always release in a finally block after the listener has been invoked. Do you see a problem with this ordering ?

jimczi commented:
> I've read this feature referred to as search contexts, reader context and point-in-time views, we should stick to one and try to use it consistently?

We settled for `search context` but are still open for suggestions. We're planning to make some renaming internally since `SearchContext` is used in many places with a different meaning (as a shard context  for a specific search phase) but we wanted to avoid polluting this PR since the change would be massive.

jimczi committed:
ensure that we remove the reader context on failures

mayya-sharipova commented on rest-api-spec/src/main/resources/rest-api-spec/api/open_search_context.json:
```diff
@@ -0,0 +1,61 @@
+    "url":{
+      "paths":[
+        {
+          "path":"/_search_context",
```
When I am trying to run `/_search_context` or `*/_search_context` or `_all/_search_contex`, I am getting this error
```
{
    "error": {
        "root_cause": [
            {
                "type": "security_exception",
                "reason": "action [indices:data/read/open_search_context] is unauthorized for user [elastic]"
            }
        ],
        "type": "security_exception",
        "reason": "action [indices:data/read/open_search_context] is unauthorized for user [elastic]",
        "caused_by": {
            "type": "illegal_state_exception",
            "reason": "There are no external requests known to support wildcards that don't support replacing their indices"
        }
    },
    "status": 403
}
```

I think it happens because `OpenSearchContextRequest` similar to `SearchRequest` should implement `implements IndicesRequest.Replaceable`, which it doesn't do right now.

dnhatn replied:
Great catch. Thanks Mayya. I've added the test and implemented this in https://github.com/elastic/elasticsearch/pull/56480/commits/52902017590b79a3325c66eef3b9e1b47467a9b9.

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
remove reader from the active list during put

dnhatn commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -621,35 +608,64 @@ public Searcher acquireSearcher(String source, SearcherScope scope) throws Engin
+                            throw new UncheckedIOException("failed to close", e);
+                        } catch (AlreadyClosedException e) {
+                            // This means there's a bug somewhere: don't suppress it
+                            throw new AssertionError(e);
```
This code has been for a long time, but happy to change to `assert false` if you prefer.

dnhatn committed:
Ensure open before acquire searcher

dnhatn committed:
check wrapper once in search supplier

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -287,16 +288,18 @@ public void afterIndexRemoved(Index index, IndexSettings indexSettings, IndexRem
         assert previous == null;
+        // ensure that if we race against afterIndexRemoved, we free the context here.
+        // this is important to ensure store can be cleaned up, in particular if the search is a scroll with a long timeout.
+        indicesService.indexServiceSafe(context.indexShard().shardId().getIndex());
```
I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/e61bd20c328276090eb84dbdf42c62d64c2a5078 to move the free logic to this method. (see https://github.com/elastic/elasticsearch/pull/56335).

dnhatn commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1158,6 +1174,21 @@ default void onFailedEngine(String reason, @Nullable Exception e) {
+
+        public final Searcher acquireSearcher(String source) {
+            final Searcher searcher = acquireSearcherInternal(source);
+            return "can_match".equals(source) ? searcher : wrapper.apply(searcher);
```
I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/e36509371e0900523c0597217d1756bb954c6c7f.

dnhatn commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1158,6 +1174,21 @@ default void onFailedEngine(String reason, @Nullable Exception e) {
+        }
+
+        public final Searcher acquireSearcher(String source) {
+            final Searcher searcher = acquireSearcherInternal(source);
```
added in https://github.com/elastic/elasticsearch/pull/56480/commits/44aca91af592aad6ddee482019091a2a8b8ee90a.

dnhatn committed:
fix test

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -403,45 +432,46 @@ public void onFailure(Exception exc) {
+        executor.execute(() -> {
+            T result;
+            try {
+                result = command.get();
```
addressed in https://github.com/elastic/elasticsearch/pull/56480/commits/0bc08e15223251b4ddb2d099f3a17222bf991a9f

dnhatn committed:
add doc for can match

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1116,14 +1149,33 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
+            Releasable releasable = readerContext.markAsUsed();
+            try {
+                final Engine.Searcher searcher = readerContext.acquireSearcher("can_match");
+                canMatchSearcher = new Engine.Searcher(searcher.source(), searcher.getDirectoryReader(),
```
We create a lightweight searcher here and use it to rewrite the query. I've added a comment in https://github.com/elastic/elasticsearch/pull/56480/commits/ae7f1a43665cee3cd1e37c6989993c2d7ea6026e.

dnhatn committed:
can_match constant

dnhatn commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1224,7 +1232,7 @@ private void markSearcherAccessed() {
         final Engine engine = getEngine();
         final Engine.Searcher searcher = engine.acquireSearcher(source, scope);
-        return wrapSearcher(searcher);
+        return "can_match".equals(source) ? searcher : wrapSearcher(searcher);
```
+1. I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/b3fe07400887d8b2bf2d598d721ece0b010e94a2.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -354,8 +354,6 @@ private static boolean searchWithCollector(SearchContext searchContext, ContextI
             }
             queryResult.searchTimedOut(true);
-        } finally {
-            searchContext.clearReleasables(SearchContext.Lifetime.COLLECTION);
```
I am wondering why we had this statement before?  I don't see anywhere in the code that we add a releasable with a scope of `Lifetime.COLLECTION`.

dnhatn replied:
Good question :). The lifetime of a SearchContext now is within a single phase or Lifetime.COLLECTION as before. With that, we do not have to track releasables with the lifetime COLLECTION separately as we can just add them to the releasables of a SearchContext.

mayya-sharipova reviewed (approved):
Thanks @dnhatn  and @jimczi . This is a great work, and so much useful.
I have reviewed the search part. I left a couple of comments to be addressed, but other than that it LGTM.

dnhatn commented:
Thanks Mayya for reviewing. I will continue addressing your comments.

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -363,37 +371,66 @@ private void loadOrExecuteQueryPhase(final ShardSearchRequest request, final Sea
+                            listener.onFailure(exc);
+                        } finally {
+                            processFailure(request, readerContext, exc);
+                        }
```
@jimczi I think we should terminate here after handling an IOException?

jimczi replied:
Good catch, I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/05dd518ed1994848e43e126434ba5434a9277294

jimczi committed:
stop execution on query rewrite failures

dnhatn committed:
Merge branch 'master' into feature/reader-context

# Conflicts:
#	client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java
#	x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/SecuritySearchOperationListenerTests.java

dnhatn committed:
add javadocs for RescoreDocIds

jimczi committed:
add AliasFilter in the search context id and rename SearchContextId into ShardSearchContextId

dnhatn committed:
fix doc test

dnhatn committed:
Add response sample for close api

dnhatn committed:
support wildcard

dnhatn commented:
@mayya-sharipova I think we have addressed your comments. Would you please take another look?

jimczi committed:
Merge branch 'master' into reader-context

dnhatn committed:
Merge branch 'master' into feature/reader-context

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/RescoreDocIds.java:
```diff
@@ -27,6 +27,11 @@
+ * Since {@link org.elasticsearch.search.internal.SearchContext} no longer hold the states of search, the top K results
+ * (i.e., documents that will be rescored by query rescorers) need to be serialized/ deserialized between search phases.
+ * A {@link RescoreDocIds} encapsulates the top K results for each rescorer by its ordinal index.
+ */
```
great javaDoc!

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/SearchContextId.java:
```diff
@@ -0,0 +1,122 @@
+            throw new IllegalArgumentException("invalid id: [" + id + "]", e);
+        }
+        try (StreamInput in = new ByteBufferStreamInput(byteBuffer)) {
+            Version version = Version.fromId(in.readVInt());
```
Do we intend to use this `version` parameter as well?

dnhatn replied:
Yes. With the version, we can add new features to the search context id without breaking BWC.

mayya-sharipova commented:
@dnhatn @jimczi  Thanks for addressing my comments. 

I was trying to test recent changes with alias filters, and I am getting the following error:

`"can't read named writeable from StreamInput"`
```
java.lang.UnsupportedOperationException: can't read named writeable from StreamInput
        at org.elasticsearch.common.io.stream.StreamInput.readNamedWriteable(StreamInput.java:1121) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.io.stream.StreamInput.readOptionalNamedWriteable(StreamInput.java:1146) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.search.internal.AliasFilter.<init>(AliasFilter.java:51) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.io.stream.StreamInput.readMap(StreamInput.java:641) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.search.SearchContextId.decode(SearchContextId.java:113) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.search.TransportSearchAction.lambda$executeRequest$2(TransportSearchAction.java:221) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
```

I think that error happens because `ByteBufferStreamInput` that is used in `SearchContextId::decode` is not `NamedWriteableAwareStreamInput` and can't read `AliasFilter::new`.

dnhatn committed:
Fix AliasFilter serialize

dnhatn commented:
> "can't read named writeable from StreamInput"

Thanks Mayya. https://github.com/elastic/elasticsearch/pull/56480/commits/4a38f0d097b1a636021eaf5e1b6b86a92ae8aa1e should address the issue.

dnhatn committed:
fix docs

dnhatn committed:
explain refresh in search context

dnhatn commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -0,0 +1,92 @@
+target indices. Sometimes it's more preferred to execute multiple search requests using
+the same point in time readers. For example, the combined result of the initial and
+subsequent search_after requests would be more consistent if they use the same point
+in time readers. This can be done by using a single search context for all search requests.
```
++. I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/e2a7fefa3fc08c40c31cddd9dec02d42f35c2362.

mayya-sharipova commented on docs/reference/search/search_context.asciidoc:
```diff
@@ -1,15 +1,16 @@
-A search context must be opened in a separate step before being used in subsequent
-search requests. The keep_alive parameter tells Elasticsearch how long it should keep
-the seach context alive, e.g. `?keep_alive=5m`.
+A search request by default executes using the most recent searchers of the target indices.
```
I really like this "point in time" notion that you had before. And also may be instead of "searchers" use "search context"?  May be a better phrasing could be:
 "A search request by default executes against the most recent point it time view of the target indices, which is called `search context`....

dnhatn replied:
++. I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/ada43b3f9cd27033143d702d6a833a594bbc7e78

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
rewording search context

mayya-sharipova commented:
@dnhatn Thanks for the latest changes, open context search on alias filter and multiple indexes works well for me! I have also checked the latest code changes, and everything LGTM!

jpountz reviewed (approved):
This PR is very large so I focused on the high level and it makes sense to me. I wonder what is the motivation for marking it beta, this looks straightforward enough to me to not feel the need to mark it beta?

jpountz commented:
Another thing is that the docs make it obvious that it works with `_search` but they don't make it obvious that it works with `_async_search` too.

dnhatn committed:
apply doc suggestion

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn committed:
bump bwc version

dnhatn committed:
Merge branch 'master' into feature/reader-context

dnhatn commented:
run elasticsearch-ci/2

dnhatn commented:
Jim and I discussed this PR offline. We decided to take a  different approach. Thanks, everyone!
